openlane/state/design_format.py

```diff
@@ -40,141 +52,181 @@ class DesignFormatObject:
-    )
-    SDF: DesignFormatObject = DesignFormatObject(
-        "sdf",
-        "sdf",
-        "Standard Delay Format",
-        multiple=True,
-    )
-    SPEF: DesignFormatObject = DesignFormatObject(
-        "spef",
-        "spef",
-        "Standard Parasitics Extraction Format",
-        multiple=True,  # nom, min, max, ...
-    )
-    LIB: DesignFormatObject = DesignFormatObject(
-        "lib",
-        "lib",
-        "LIB Timing Library Format",
-        multiple=True,
-    )
-    SPICE: DesignFormatObject = DesignFormatObject(
-        "spice",
-        "spice",
-        "Simulation Program with Integrated Circuit Emphasis",
-    )
-
-    MAG: DesignFormatObject = DesignFormatObject(
-        "mag",
-        "mag",
-        "Magic VLSI View",
-    )
-
-    GDS: DesignFormatObject = DesignFormatObject(
-        "gds",
-        "gds",
-        "GDSII Stream",
-    )
-    MAG_GDS: DesignFormatObject = DesignFormatObject(
-        "mag_gds",
-        "magic.gds",
-        "GDSII Stream (Magic)",
-    )
-    KLAYOUT_GDS: DesignFormatObject = DesignFormatObject(
-        "klayout_gds",
-        "klayout.gds",
-        "GDSII Stream (KLayout)",
+    @property
+    @deprecated(
+        "The DesignFormat is directly returned now, no need for .value",
+        version="3.0.0",
+        action="once",
     )
+    def value(self) -> DesignFormat:
+        return self
 
-    JSON_HEADER: DesignFormatObject = DesignFormatObject(
-        "json_h",
-        "h.json",
-        "Design JSON Header File",
-    )
-    VERILOG_HEADER: DesignFormatObject = DesignFormatObject(
-        "vh",
-        "vh",
-        "Verilog Header",
-    )
+    def register(self):
+        self.__class__.factory.register(self)
 
     def __str__(self) -> str:
-        return self.value.id
+        return self.id
+
+    def __hash__(self):
+        return hash(self.id)
 
     @staticmethod
+    @deprecated(
+        "Use DesignFormat.factory.get",
+        version="3.0.0",
+        action="once",
+    )
     def by_id(id: str) -> Optional["DesignFormat"]:
-        return _designformat_by_id.get(id)
-
-
-_designformat_by_id: Dict[str, "DesignFormat"] = {
-    format.value.id: format for format in DesignFormat
-}
+        return DesignFormat.factory.get(id)
+
+    class DesignFormatFactory(object):
+        """
+        A factory singleton for DesignFormats, allowing them to be registered
+        and then retrieved by a string name.
+
+        See https://en.wikipedia.org/wiki/Factory_(object-oriented_programming) for
+        a primer.
+        """
+
+        _registry: ClassVar[Dict[str, DesignFormat]] = {}
+
+        @classmethod
+        def register(Self, df: DesignFormat) -> DesignFormat:
+            """
+            Adds a DesignFormat to the registry using its
+            :attr:`DesignFormat.id`, :attr:`DesignFormat.name` and
+            :attr:`DesignFormat.alts` attributes.
+            """
+            Self._registry[df.id] = df
```

Comment:
does this mean that `id`, `name` and `alts` all should be unique to each `DesignFormat`

Reply:
yes. just like step

Reply:
will openlane cause exception if attempting to register a `DesignFormat` with a duplicate of any of the above keys